src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -58,27 +58,43 @@ impl HummockSnapshotManager {
     }
 
     pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> Result<()> {
-        let local_count = async {
+        let min_epoch = async {
+            // Decrease the ref count of corresponding epoch. If all last pinned epoch is dropped,
+            // mark as outdated.
             let mut core_guard = self.core.lock().await;
             let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
             if let Some(query_ids) = query_ids {
                 query_ids.remove(query_id);
+                if query_ids.is_empty() && epoch == core_guard.last_pinned {
+                    core_guard.is_outdated = true;
+                }
+            }
+
+            // Iterate for all query and calculate the min epoch.
+            let mut min_epoch = None;
+            for (epoch, query_ids) in &core_guard.epoch_to_query_ids {
                 if query_ids.is_empty() {
-                    core_guard.epoch_to_query_ids.remove(&epoch);
-                    if epoch == core_guard.last_pinned {
-                        core_guard.is_outdated = true;
-                    }
-                    return true;
+                    min_epoch = Some(*epoch);
+                    break;
                 }
             }
-            false
+
+            if let Some(epoch_inner) = min_epoch {
+                core_guard.epoch_to_query_ids.remove(&epoch_inner);
+                assert!(epoch_inner <= core_guard.last_pinned);
+            }
+            min_epoch
         };
-        let need_to_request_meta = local_count.await;
-        if need_to_request_meta {
+
+        let need_to_request_meta = min_epoch.await;
+        if let Some(epoch_inner) = need_to_request_meta {
+            tracing::info!("Unpin epoch in frontend manager {:?}", epoch_inner);
             let meta_client = self.meta_client.clone();
             tokio::spawn(async move {
-                let handle = tokio::spawn(async move { meta_client.unpin_snapshot(epoch).await });
-
+                let handle =
+                    tokio::spawn(
+                        async move { meta_client.unpin_snapshot_before(epoch_inner).await },
+                    );
```

Comment:
We don't have error handling here? If the `unpin_snapshot_before` rpc call failed, we may leak snapshot.

Reply:
I have considered this part.

If `unpin_snapshot_before` failed, the snapshot is indeed leaked (FE do not aware that it pins the epoch). But if next `unpin_snapshot_before` succeed , there is not actual leak cuz the next RPC call will help the previous failed RPC.

Reply:
If we remove the unpinned epoch in FE iff the RPC is succeed, I'm concerned that we might send a lot of same unpin RPC to meta (They all tries to unpin the same epoch). 

I think better do this after we changed our unpin timer strategy, e.g., unpin only after same events happened.

Reply:
But it increases the time of maintaining snapshot. I think it would be better to add a retry here?

Reply:
Agreed. Error control should be considered in future while not in this PR.

Reply:
OK, but please output some logs for the result of unpin, so that at least we can know whether the result of unpin in trouble shooting.

Reply:
I agree that we should have better observability, but I still suggest to add some log as a short term diagnose method.

Reply:
Added some logs. Suggestions welcomed.